src/grpcbox_socket.erl

```diff
@@ -11,14 +11,67 @@
          code_change/3,
          terminate/2]).
 
-%% public api
+-record(state, {
+    pool,
+    listen_opts,
+    pool_opts,
+    socket,
+    mref
+}).
 
+%% public api
 start_link(Pool, ListenOpts, AcceptorOpts) ->
     gen_server:start_link(?MODULE, [Pool, ListenOpts, AcceptorOpts], []).
 
 %% gen_server api
 
 init([Pool, ListenOpts, PoolOpts]) ->
+    io:format("starting ~p", [?MODULE]),
+    {ok, #state{pool = Pool, pool_opts = PoolOpts, listen_opts = ListenOpts}, 0}.
+
+handle_call(Req, _, State) ->
+    {stop, {bad_call, Req}, State}.
+
+handle_cast(Req, State) ->
+    {stop, {bad_cast, Req}, State}.
+handle_info(timeout, State) ->
+    case start_listener(State) of
+        {ok, {Socket, MRef}} ->
+            {noreply, State#state{socket = Socket, mref = MRef}};
+        _ ->
+            erlang:send_after(5000, self(), init),
+            {noreply, State}
+    end;
+handle_info({'DOWN', MRef, port, Socket, _Reason}, #state{mref = MRef, socket = Socket} = State) ->
+    io:format("got down from socket, restarting"),
+    catch gen_tcp:close(Socket),
+    erlang:send_after(5000, self(), init),
```

Comment:
this atom that we're self-sending should be `timeout` no?

Reply:
indeed, good catch